cg/meta/demultiplex/utils.py

```diff
@@ -35,27 +36,32 @@ def get_lane_from_sample_fastq(sample_fastq_path: Path) -> int:
     raise ValueError(f"Could not extract lane number from fastq file name {sample_fastq_path.name}")
 
 
-def get_sample_id_from_sample_fastq(sample_fastq: Path) -> str:
-    """
-    Extract sample id from fastq file path.
+def get_sample_fastqs_from_flow_cell(
+    flow_cell_directory: Path, sample_internal_id: str
+) -> Optional[List[Path]]:
+    """Retrieve sample FastQs from a flow cell directory."""
+    root_pattern = f"{sample_internal_id}_S*_L*_R*_*{FileExtensions.FASTQ}{FileExtensions.GZIP}"
+    unaligned_pattern = f"Unaligned*/Project_*/Sample_{sample_internal_id}/*{FileExtensions.FASTQ}{FileExtensions.GZIP}"
 
-    Pre-condition:
-        - The parent directory name contains the sample id formatted as Sample_<sample_id>
-    """
+    for pattern in [root_pattern, unaligned_pattern]:
+        sample_fastqs: List[Path] = get_files_matching_pattern(
+            directory=flow_cell_directory, pattern=pattern
+        )
+        valid_sample_fastqs: List[Path] = get_valid_sample_fastqs(
+            fastq_paths=sample_fastqs, sample_internal_id=sample_internal_id
+        )
 
-    sample_parent_match = re.search(r"Sample_(\w+)", sample_fastq.parent.name)
+        if valid_sample_fastqs:
+            return valid_sample_fastqs
 
-    if sample_parent_match:
-        return sample_parent_match.group(1)
-    else:
-        raise ValueError("Parent directory name does not contain 'Sample_<sample_id>'.")
 
-
-def get_sample_fastq_paths_from_flow_cell(flow_cell_directory: Path) -> List[Path]:
-    fastq_sample_pattern: str = (
-        f"Unaligned*/Project_*/Sample_*/*{FileExtensions.FASTQ}{FileExtensions.GZIP}"
-    )
-    return list(flow_cell_directory.glob(fastq_sample_pattern))
+def get_valid_sample_fastqs(fastq_paths: List[Path], sample_internal_id: str) -> List[Path]:
+    """Return a list of valid fastq files."""
+    return [
+        fastq
+        for fastq in fastq_paths
+        if is_valid_sample_fastq_file(sample_fastq=fastq, sample_internal_id=sample_internal_id)
+    ]
```

Comment:
```suggestion
def get_valid_fastq_file_paths(fastq_file_paths: List[Path], sample_internal_id: str) -> List[Path]:
    """Return a list of valid fastq files."""
    return [
        fastq_file_path
        for fastq_file_path in fastq_file_paths
        if is_valid_fastq_file(fastq_file_path=fastq_file_path, sample_internal_id=sample_internal_id)
    ]
```
I think it might be better to be explicit that we are passing file paths here in the code.

Reply:
I don't agree necessarily. I think the current naming of the function is clear enough and I will keep it as is.